### ISSUE TITLE
How to embed asynchronous pipelines in TaskFlow

### ISSUE THREAD

**Issue by kzuiderveld** [taskflow/taskflow#426]
I'm trying to use TaskFlow for an application that contains a mix of GPU and CPU-based processing; the caveat is that the GPU processing is not using a local GPU/CUDA but is using an NVIDIA Triton inference server (for deep learning). 

So, as part of the computational graph, we need to do inference using a Triton server; as it will take a while for the data to be uploaded, inference being done, and results being retrieved, we would like to do the inference asynchronously (i.e. the worker thread will steal work while waiting for the inference data to arrive). 

I can't seem to figure out how to implement this using the various constructs - largely because I'm still climbing the TF learning curve. Tried async from the executor, define a Subflow -  they don't quite do what I need.

So, basically, if I'm executing a task - how can I yield back control to the executor while I'm waiting for data to arrive - and once the data arrives, how can I resume the work? (hmm, that does sound like coroutines, doesn't it)

Question 2: Inference is faster when we combine data into batches. Would there be a way to implement this batching using TF?

Thanks for the great work!

Karel

**Comment by tsung-wei-huang** [comment]
Hi @kzuiderveld thank you for using taskflow in your deep-learning workload. Your solution idea is correct - i.e. the worker thread will steal work while waiting for the inference data to arrive. This can be done using the new [tf::Executor::loop_until](https://taskflow.github.io/taskflow/classtf_1_1Executor.html#afbff30c324c5887aace09e7fe9a0d607) which prevents the caller thread from blocking but put that worker into the work-stealing loop. This method should only be called by a worker of an executor.

```cpp
taskflow.emplace([&](){
  std::cout << "submitting work to GPU...\n";
  run_inference_on_gpu();
  executor.loop_until([](){     // can only be called from an worker of an executor 
                                               // the caller worker is not blocking but joins the work-stealing loop of the executor
    if(inference_done()) return true;
    else return false;
  });
});
```

In practice, the function `inference_done()` can be something like polling if a CUDA event has been completed, but keep in mind calling the CUDA event too frequently will give you some runtime costs.

Currently, `tf::Executor::loop_until` is only available in the master branch and will be merged to the next release.

**Comment by tsung-wei-huang** [comment]
close this issue first - feel free to reopen it if you have additional questions

**Comment by albertchristianto** [comment]
Hi @tsung-wei-huang,

I am very interested with this topic. I am currently working on the streaming project for deep learning, which has multiple deep learning models in multiple stages. I have read your example project about the text processing pipeline. However I have some problems to understand and adjust the project from the text processing pipeline example, therefore I want to ask your opinion about it.

Here is the Flowchart for the streaming project for deep learning. (I divide the flowchart into stages per iteration of an image from the streaming video)
1. Get an image, pre-process the image.
2. text detection, inference in GPU
3. post-processing the results, pre-process the image based on the results
4. text recognition, inference in GPU
5. post-processing the results

for the steps no 2 and 4, I will have N deep learning models which can be spawn during the initialization (I suppose the deep learning model will be provided to the taskflow worker using object pointer(?))  

best regards, 
Albert
